### ISSUE TITLE
How2Sign Inference Not Working

### ISSUE THREAD

**Issue by itsjustfons** [HenryLittle/GloFE#5]
I'm currently trying to run the Inference segment on the How2Sign dataset. The pose data given by how2sign is in the form of a json file. The README of GloFE also says to arrange the json files in a certain way. The train_how2_pose_DDP_inter_VN.py however only reads pose data as a pkl file.

TL:DR. How2sign pose data is in json. GloFE reads pose data of How2sign by looking for pkl files

Is there an update which has not been uploaded?

**Comment by SummitStudiosDev** [comment]
@itsjustfons Did you ever find the pose data? I'm looking for it here as well.

**Comment by itsjustfons** [comment]
Still no solution on this unfortunately

**Comment by moaldeen** [comment]
@itsjustfons I am looking to run any model that allows me to translate a video to text ,,, is there anything i can do without spending so much time on ?

**Comment by Vaibhavmulti** [comment]
> I'm currently trying to run the Inference segment on the How2Sign dataset. The pose data given by how2sign is in the form of a json file. The README of GloFE also says to arrange the json files in a certain way. The train_how2_pose_DDP_inter_VN.py however only reads pose data as a pkl file.
> 
> TL:DR. How2sign pose data is in json. GloFE reads pose data of How2sign by looking for pkl files
> 
> Is there an update which has not been uploaded?

@itsjustfons Did you find the solution for this?

@HenryLittle 
I got the B-F-H 2d keypoint clips from  https://how2sign.github.io/index.html#download
The format of the data is 
/test/openpose_output/json
/test/openpose_output/videos

![image](https://github.com/user-attachments/assets/d32b4da4-2798-4fcf-916b-f95426b4d7b4)

Inside the JSON there are multiple folders each having multiple JSONS.
It appears consistent with what has been asked for in the readme.

![image](https://github.com/user-attachments/assets/c3cfd650-8af1-4f2d-8854-a62f4ec6986f)

However, when loading the dataset the Python script reads the .pkl file. 

Have you converted the entire files into 1 pickle? Or are you getting the pickle from somewhere else?

**Comment by HenryLittle** [comment]
@Vaibhavmulti Yes, the multiple json files are combined into one .pkl file to speed up the dataset loading time. The shape is just [Time, Vertices, Channels].

**Comment by Vaibhavmulti** [comment]
> @Vaibhavmulti Yes, the multiple json files are combined into one .pkl file to speed up the dataset loading time. The shape is just [Time, Vertices, Channels].

Thanks @HenryLittle  for the quick reply.
So for example _fZbAxSSbX4_0-5-rgb_front

_fZbAxSSbX4_0-5-rgb_front_000000000000_keypoints.json
_fZbAxSSbX4_0-5-rgb_front_000000000001_keypoints.json


...
Are the timeframes (frames in the video) Making the first dimension?


And have you flattened pose_keypoints_2d, face_keypoints_2d, hand_left_keypoints_2d, hand_right_keypoints_2d into a single vector making the 2nd dimension?

And the 3rd dimension is the x,y,confidence score?

**Comment by HenryLittle** [comment]
@Vaibhavmulti It should be the case. As for the order of different keypoints, you can refer to the code to be sure. If I remembered it correctly, the order should be [body, hand_l, hand_r, face].

**Comment by Vaibhavmulti** [comment]
@HenryLittle  
Another thing in the How2SignPoseDataset class you are filtering out the missing parts.

## filter out missing parts
with open('/mnt/user/E-linkezhou.lkz-385206/workspace/How2Sign/tools/how2sign_missing.txt', 'r') as f:
    names = f.readlines()
    missing = [x.strip() for x in names]
    
You have hardcoded the missing.txt file path , I can't seem to find this file anywhere.

**Comment by HenryLittle** [comment]
@Vaibhavmulti You should be able to generate the file yourself by cross checking the downloaded videos. If there is no missing ones, you can just delete this in code.

**Comment by Vaibhavmulti** [comment]
![image](https://github.com/user-attachments/assets/2d370ca7-2dc2-468e-b6a9-1c921bfaee3c)
@HenryLittle I ran your inference code on the test split only. 
I got the following results. Ignore the val part.

While the paper mentions 
![image](https://github.com/user-attachments/assets/e7a63e6f-4dea-4bd0-9ccd-fcda5461968b)

Am I missing something?

**Comment by HenryLittle** [comment]
@Vaibhavmulti You can cross check the detailed generated output with the ones included with the ckpt file. That might give you some clue about what's wrong. Unfortunately, I don't have the access to the enterprise server anymore. So I couldn't provide a more detailed guide.

**Comment by Vaibhavmulti** [comment]
@HenryLittle 
I had ordered the key points incorrectly. Now I am getting this as the score.
![image](https://github.com/user-attachments/assets/b49c3018-dd72-4580-953a-cfccff00ef26)
It is close to the one mentioned in the paper.

Is this okay or should I exactly aim for the ones mentioned in the paper?

**Comment by HenryLittle** [comment]
@Vaibhavmulti If you want to to get closer to the numbers in the paper, I would suggest you use an older version of the code without the randomness fix. However, if you want to proceed with the code, I would suggest you retrain the model based on the current version of the code.

**Comment by itsjustfons** [comment]
Thanks for the support so far on this thread. I need some help with formatting the pkl files into what the "read_pose_files" wants. Is there a code you used to convert the json pose data into pkl?
@Vaibhavmulti @HenryLittle

**Comment by itsjustfons** [comment]
Worked it out.
Notes for others:
The other json files within the same subfolder would correspond to other frames of the same video. 


Format for pkl files: [Time, Vertices, Channels], 
Time: Corresponds to the frames. Each json file contains 1 frame worth of data
Vertices: A flattened version of the pose keypoint values. I only need to work with the 2d data. The flattened data should follow the order: [body, hand_l, hand_r, face].
Channel: There are 3 channels: x coordinate, y coordinate, and confidence value

**Comment by Analroy** [comment]
## Request for Pickle File Creation Guide

Hi,

Can you please provide a detailed guide on how to create those pickle files?

To clarify, could you confirm the following:

- **T = 120** (Number of frames)
- **V** = A flattened vector consisting of:
  - 74 body keypoints
  - 62 keypoints for hand 1
  - 62 keypoints for hand 2
  - 209 face keypoints  
  Or have you used a different number of keypoints in each case?

- **C** = 3 channels:
  - x-coordinate
  - y-coordinate
  - Confidence value  

Additionally, where can I find this information? The corresponding `.json` file does not seem to contain these fields.

Thanks!
